fractional-NFT/project/token-distributor-contract/src/main.sw

```diff
@@ -0,0 +1,251 @@
+dep interface;
+dep utils;
+
+use data_structures::{DistributionState, TokenDistribution};
+use errors::{AccessError, AssetError};
+use events::{
+    BuybackEvent,
+    CancelEvent,
+    CloseEvent,
+    CreateEvent,
+    PurchaseAdminEvent,
+    PurchaseEvent,
+    ReserveEvent,
+    SellEvent,
+    TokenPriceEvent,
+    WithdrawEvent,
+};
+use interface::TokenDistributor;
+use std::{
+    auth::msg_sender,
+    call_frames::{
+        contract_id,
+        msg_asset_id,
+    },
+    context::{
+        msg_amount,
+        this_balance,
+    },
+    logging::log,
+    token::transfer,
+};
+use utils::{create_fractional_nft, fractional_nft_supply, withdraw_fractional_nft};
+
+storage {
+    /// Maintains a mapping of information on every token distribution started using this contract.
+    /// Mapping(Token Contract -> Token Distribution Information)
+    token_distributions: StorageMap<ContractId, Option<TokenDistribution>> = StorageMap {},
+}
+
+impl TokenDistributor for Contract {
+    #[storage(read, write)]
+    fn buyback(fractional_nft: ContractId, token_price: u64) {
+        require(storage.token_distributions.get(fractional_nft).is_some(), AccessError::DistributionDoesNotExist);
+        let mut token_distribution = storage.token_distributions.get(fractional_nft).unwrap();
+
+        require(token_distribution.admin.is_some() && token_distribution.admin.unwrap() == msg_sender().unwrap(), AccessError::NotTokenAdmin);
+        require(token_distribution.state == DistributionState::Distributed, AccessError::InvalidState);
+
+        // Ensure the seller is only providing the amount to buyback the number of tokens that have been sold
+        let token_supply = fractional_nft_supply(fractional_nft);
+        require(msg_amount() == (token_supply - this_balance(fractional_nft)) * token_price && msg_asset_id() == token_distribution.external_asset, AssetError::InvalidAssetTransfer);
+
+        // Store the buyback price and change the state
+        token_distribution.token_price = token_price;
+        token_distribution.state = DistributionState::Buyback;
+        storage.token_distributions.insert(fractional_nft, Option::Some(token_distribution));
+
+        log(BuybackEvent {
+            fractional_nft,
+            token_price,
+        });
+    }
+
+    #[storage(read, write)]
+    fn create(
+        external_asset: ContractId,
+        fractional_nft: ContractId,
+        nft: ContractId,
+        reserve_price: Option<u64>,
+        token_admin: Option<Identity>,
+        token_price: u64,
+        token_supply: u64,
+        nft_token_id: u64,
+    ) {
+        require(storage.token_distributions.get(fractional_nft).is_none(), AccessError::DistributionAlreadyExists);
+
+        match token_admin {
+            Option::Some(admin) => {
+                create_fractional_nft(Option::Some(Identity::ContractId(contract_id())), fractional_nft, nft, token_supply, nft_token_id);
+            },
+            Option::None(admin) => {
```

Comment:
```suggestion
            Option::None => {
```

None doesn't take params, before accepting here check in the code to see if compiles or if the bug persists in the compiler